### ISSUE TITLE
DTDL-Generalize messages with message types and related schemas

### ISSUE THREAD

**Issue by jonmikeli** [Azure/IoTPlugandPlay#43]
Hi,

Is there any plan to generalize message types and replace the telemetry type by a generic message structure with, for instance:
 - messageType (ex: telemetry, error, system, etc)
 - messageSchema
 - messageSchemaVersion
and allowing a device (or module) to send different types of messages.

Unless I have misunderstood something, this is not possible today.
I have been confronted with this need in different projects and it is actually quite useful.

**Comment by jonmikeli** [comment]
I had missed the `object` type in the `schema` property of the `telemetry` type.


![image](https://user-images.githubusercontent.com/22737122/71590682-bc3d2a00-2b29-11ea-8a00-88b37bd9b5b2.png)


`schema` property values:

![image](https://user-images.githubusercontent.com/22737122/71590560-4cc73a80-2b29-11ea-9835-03ec049012e3.png)

`object` type:

![image](https://user-images.githubusercontent.com/22737122/71590596-74b69e00-2b29-11ea-9e89-8ac3356b51a5.png)

An example:

![image](https://user-images.githubusercontent.com/22737122/71590840-44bbca80-2b2a-11ea-910b-460bdd16bff7.png)

Source:
https://github.com/Azure/IoTPlugandPlay/tree/master/DTDL

I guess this should allow sending "messages" of different types through telemetry.

Any plan to add JSON schema references to the `object` type (ideally with the schema version)?

**Comment by briancr-ms** [comment]
Yes, the intention of DTDL's telemetry is to enable messages of nearly any type to be defined. These could be telemetry messages that contain measurements sent every minute, like temperature and humidity. These could be telemetry messages that are sent only when an event occurs, like when a door opens or an error occurs.

We hadn't planned on adding JSON Schema support, but can you elaborate more on what you're thinking, perhaps with an example?

**Comment by jonmikeli** [comment]
Thanks @briancr-ms.
Sure, no problem to elaborate.

The idea behind my question was to control the structure of incoming messages and only accept in the system the valid ones (in terms of schema and allowed values, if specified in the schema).
It adds consistency to the whole solution and minimizes the data being stored and going through the processes in the solution.

I have coded that feature with Azure Functions in many projects and it surprised me how useful it has been. If you need code examples I can send you a few (I am also cleaning a private repository to publish it and it will contain the mentioned feature with many others).

**Comment by briancr-ms** [comment]
I understand the value of validating incoming messages based on the DTDL telemetry definitions. The intention of DTDL is to capture enough information so that validators like you suggest can be built. In general, it should be possible to generate a JSON Schema from a DTDL definition.

**Comment by ahernandez-allegion** [comment]
@jonmikeli @briancr-ms 

I completely agree with this initiative. I believe _Telemetry_ should be a semantic variation on the fundamental capability of a _Message_. Furthermore, _Event_ should be another semantic type of a _Message_ capability.

There are huge fundamental differences between Telemetry, Events, and Messages. Telemetry can be dropped. Events cannot. An _Event_ can be _Telemetry_. _Telemetry_ is not necessarily an _Event_.

There are also significant temporal differences around timestamping, which is also what differentiates a basic _Message_ from the other two (_Messages_ don't necessarily have a time component).

Replacing _Telemetry_ with a more fundamental _Message_ (or _Event_) type and adding _Telemetry_ and _Event_ would also be more inline with other DSLs:

https://www.w3.org/TR/wot-thing-description/#thing
https://www.eclipse.org/ditto/basic-feature.html
https://webthings.io/schemas/

**Comment by jonmikeli** [comment]
Hi @jrdouceur ,

Thanks for the update. I read about the Parsing/Validation feature some time ago, which is great. At that time, DTDL v2 had some limitations in terms of what could be defined (and how). I need to assess again all this with [DTDL v3](https://azure.github.io/opendigitaltwins-dtdl/DTDL/v3/DTDL.v3.html).
